File: old/source/frontend/scanner.cpp
#include <frontend/scanner.hpp>

#include <unordered_map>

// Keywords
static std::unordered_map<string, TokenKind> keywords =
{
    { "export", TokenKind::Export },
    { "import", TokenKind::Import },

    { "return", TokenKind::Return },
};

// Set the source
void Scanner::set(string _source)
{
    source = _source;
    start  = 0;
    end    = 0;
    line   = 1;
    column = 1;
}

// Is character an identifier?
static bool is_identifier(char character)
{
    return (character >= 'A' && character <= 'Z') || 
           (character >= 'a' && character <= 'z') ||
           (character == '_');
}

// Is character a number?
static bool is_number(char character)
{
    return (character >= '0' && character <= '9');
}

// Advance the scanner
char Scanner::advance(void)
{
    ++column;
    return source[end++];
}

// Skip whitespace
void Scanner::skip_whitespace(void)
{
    for (;;)
    {
        switch (source[end])
        {
            case '\n':
            {
                line   += 1;
                column  = 0;
            }

            case ' ':
            case '\t':
            case '\r':
            {
                advance();
                break;
            }

            default:
                return;
        }
    }
}

// Make token
TokenKind Scanner::make_token(TokenKind kind, TokenData data)
{
    current.kind   = kind;
    current.data   = data;
    current.line   = line;
    current.column = column;

    return kind;
}

TokenKind Scanner::make_token(TokenKind kind)
{
    current.kind   = kind;
    current.data   = std::string(source).substr(start, end - start).c_str();
    current.line   = line;
    current.column = column;

    return kind;
}

// Make error token
TokenKind Scanner::make_token(string message)
{
    current.kind   = TokenKind::Error;
    current.data   = message;
    current.line   = line;
    current.column = column;

    return TokenKind::Error;
}

// Make string token
TokenKind Scanner::make_string_token(void)
{
    while (source[end] != '"' && source[end] != '\0')
        advance();

    if (source[end] == '\0')
        return make_token("Unterminated string.");

    advance();
    return make_token(TokenKind::String);
}

// Make identifier token
TokenKind Scanner::make_identifier_token(void)
{
    while (is_identifier(source[end]) || is_number(source[end]))
        advance();

    // Find keyword
    std::string keyword = std::string(source).substr(start, end - start);

    if (keywords.find(keyword.c_str()) != keywords.end())
        return make_token(keywords[keyword.c_str()], keyword.c_str());

    return make_token(TokenKind::Identifier, keyword.c_str());
}

// Make number
TokenKind Scanner::make_number_token(void)
{
    while (is_number(source[end]))
        advance();

    if (source[end] == '.')
    {
        advance();

        while (is_number(source[end]))
            advance();
    }

    return make_token(TokenKind::Number, std::stod(source + start));
}

// Scan token
TokenKind Scanner::scan(void)
{
    skip_whitespace();

    start    = end;
    previous = current;

    if (!source[end])
        return make_token(TokenKind::End, "<EOF>");

    int8 character = advance();

    if (is_identifier(character))
        return make_identifier_token();

    if (is_number(character))
        return make_number_token();

    switch (character)
    {
        case '(':
            return make_token(TokenKind::LeftParenthesis);

        case ')':
            return make_token(TokenKind::RightParenthesis);

        case '{':
            return make_token(TokenKind::LeftBrace);

        case '}':
            return make_token(TokenKind::RightBrace);

        case ',':
            return make_token(TokenKind::Comma);

        case ';':
            return make_token(TokenKind::Semicolon);

        case '.':
            return make_token(TokenKind::Dot);

        case '"':
            return make_string_token();
    }

    return make_token("Unexpected character.");
}

File: source/frontend/Scanner.cs
using System;
using System.Collections.Generic;

namespace Suteki
{
    class Scanner
    {
        private int    Line;
        private int    Column;
        private int    Start;
        private int    End;
        private string Source;
        public  Token  Previous;
        public  Token  Current;

        private Dictionary<string, TokenKind> Keywords = new Dictionary<string, TokenKind>
        {
            { "export", TokenKind.Export },
            { "import", TokenKind.Import },

            { "extern", TokenKind.Extern },

            { "true",  TokenKind.Bool    },
            { "false", TokenKind.Bool    },

            { "return", TokenKind.Return },
        };

        // Set scanner input
        public void Set(string source)
        {
            Line     = 0;
            Column   = 0;
            Start    = 0;
            End      = 0;
            Source   = source;
            Previous = null;
            Current  = null;
        }

        // Advance character
        private char Advance()
        {
            ++End;
            ++Column;

            return Source[End - 1];
        }

        // Match current character?
        private bool Match(char expected)
        {
            if (End < Source.Length && Source[End] == expected)
            {
                Advance();
                return true;
            }

            return false;
        }

        // Character matches identifier?
        private bool IsIdentifier(char character)
        {
            return (character >= 'A' && character <= 'Z') ||
                   (character >= 'a' && character <= 'z') ||
                   (character == '_');
        }

        // Character matches number?
        private bool IsNumber(char character)
        {
            return (character >= '0' && character <= '9');
        }

        // Skip whitespace
        private void SkipWhitespace()
        {
            for (;;)
            {
                // Is source end?
                if (End >= Source.Length)
                    return;

                switch (Source[End])
                {
                    case '\n':
                    {
                        Advance();

                        Line   += 1;
                        Column  = 0;

                        break;
                    }

                    case ' ' :
                    case '\r':
                    case '\t':
                    {
                        Advance();
                        break;
                    }

                    default:
                        return;
                }
            }
        }

        // Make token
        private TokenKind MakeToken(TokenKind kind)
        {
            Current = new Token
            { 
                Kind   = kind,
                Data   = Source.Substring(Start, (End - Start)),
                Line   = Line,
                Column = Column,
            };

            return kind;
        }

        // Make token
        private TokenKind MakeToken(TokenKind kind, object data)
        {
            Current = new Token
            { 
                Kind   = kind,
                Data   = data,
                Line   = Line,
                Column = Column,
            };

            return kind;
        }

        // Make error token
        private TokenKind MakeToken(string message)
        {
            Current = new Token
            { 
                Kind   = TokenKind.Error,
                Data   = message,
                Line   = Line,
                Column = Column,
            };

            return TokenKind.Error;
        }

        // Make number token
        private TokenKind MakeNumberToken()
        {
            bool isFloat = false;

            while (End < Source.Length && IsNumber(Source[End]))
                Advance();

            if (Match('.'))
            {
                isFloat = true;

                while (End < Source.Length && IsNumber(Source[End]))
                    Advance();
            }

            string numberString = Source.Substring(Start, (End - Start));
            return MakeToken((isFloat) ? TokenKind.Float : TokenKind.Integer, double.Parse(numberString));
        }

        // Make identifier token
        private TokenKind MakeIdentifierToken()
        {
            while (End < Source.Length && IsIdentifier(Source[End]))
                Advance();

            string identifierString = Source.Substring(Start, (End - Start));

            if (Keywords.ContainsKey(identifierString))
                return MakeToken(Keywords[identifierString]);

            return MakeToken(TokenKind.Identifier, identifierString);
        }

        // Make string token
        private TokenKind MakeStringToken()
        {
            while (End < Source.Length && Source[End] != '"')
                Advance();

            if (End >= Source.Length)
                return MakeToken("Unterminated string.");

            Advance();
            return MakeToken(TokenKind.String);
        }

        // Scan token
        public TokenKind Next()
        {
            SkipWhitespace();

            Previous = Current;
            Start    = End;

            // Is source end?
            if (End >= Source.Length)
                return MakeToken(TokenKind.End, "<EOF>");

            char character = Advance();

            if (IsNumber(character))
                return MakeNumberToken();

            if (IsIdentifier(character))
                return MakeIdentifierToken();

            switch (character)
            {
                case '(':
                    return MakeToken(TokenKind.LeftParenthesis);

                case ')':
                    return MakeToken(TokenKind.RightParenthesis);

                case '{':
                    return MakeToken(TokenKind.LeftBrace);

                case '}':
                    return MakeToken(TokenKind.RightBrace);

                case ',':
                    return MakeToken(TokenKind.Comma);

                case ';':
                    return MakeToken(TokenKind.Semicolon);

                case '.':
                    return MakeToken(TokenKind.Dot);

                case '"':
                    return MakeStringToken();

                default:
                    return MakeToken("Unexpected character.");
            }
        }
    }
}